### ISSUE TITLE
Single graph but different node features

### ISSUE THREAD

**Issue by shingyipcheung** [pyg-team/pytorch_geometric#857]
## ❓ Questions & Help

First of all, thanks `rusty1s` for creating this amazing GNN library.

I have a dataset which contains a static graph but many instances of different node features, that is having the same `edge_index` but different  `x` states of the same node.

Previously, I used `networkx` and ordinary `torch` to recursively propagate the node information for "message passing". In the training phase, for each batch, there is a reused directed graph object and each node has a batch of `x` of size `(batch_size, num_features)`. Finally, I concatenate all the node embeddings `(batch_size, num_nodes * out_channel)` to encode whole graph states `(batch_size, entire_graph_out_channel)` for some classification tasks.

While in `pytorch_geometric`,  the input `x` of an operator `conv(x, edge_index)` is of size `(num_nodes, num_features)`. Are there examples or classes I need to look at for this case?

**Comment by rusty1s** [comment]
Hi, there is no example yet but I recently added support for static graphs. You can look it up [here](https://github.com/rusty1s/pytorch_geometric/issues/817#issuecomment-556967316). All you need to do is reshape `x` into format `[batch_size, num_nodes, num_features]` and set the `node_dim` attribute for all message passing layers.

**Comment by shingyipcheung** [comment]
Thanks. I've tried to pull the repo and install torch geometric again. It works now. 

Can it be also applied to the pooling layers?

**Comment by rusty1s** [comment]
This depends on your choice of pooling layer. If it does not depend on node features, you can just apply it to your single graph and proceed with your coarsened graph.

**Comment by ziye3001** [comment]
Hi @rusty1s , I just followed your instruction and it works very well for convolution and pooling layers. However, when I was trying to add BatchNorm1D, it seems that pytorch only supports N * C * L, where N is the batch dimension and C is the feature dimension. So I have to permute the dimensions before the batch normalization and permute them back afterwards? Thank you!

**Comment by rusty1s** [comment]
You are right, you might need to reshape your tensor first to `[batch_size * num_nodes, num_features]`.

**Comment by engmubarak48** [comment]
@rusty1s In pytorch, the models support n dimensional tensors and only rely on the last dimension. For example if my `x = (N, T, F)` and I have my edge_index as normal.. how can I apply it to the geometric models like GCN? do i need to reshape to (N*T, F).. because the message passing should Just update F to some D and N and T stay the same?

**Comment by rusty1s** [comment]
Yes, you can do this. However, not all GNN layers in PyG support three-dimensional input features. `GCNConv` does support it though. See the "static" column [here](https://pytorch-geometric.readthedocs.io/en/latest/cheatsheet/gnn_cheatsheet.html).

**Comment by engmubarak48** [comment]
Based on the Docs:

`
static: If checked (✓), supports message passing in static graphs, e.g., GCNConv(...).forward(x, edge_index) with x having shape [batch_size, num_nodes, in_channels].
`

But in my case, the shape is more like [N, T, in_channels] where T is like the time feature and N is batch_size * num_nodes (like you would have in normal Data object). In other words, my nodes are still part of the batch. And I want my message passing model to update the `in_channels` to some `out_channels`. 

I think I understand that this feature (`static`) might be doing what I want but I want to clarify that my input node features are not [batch_size, num_nodes, in_channels] but [batch_size * num_nodes, T, in_channels].

**Comment by engmubarak48** [comment]
@rusty1s we could do this but it is not optimized.
`x = torch.stack([self.gnn(x[:, t, :], graph.edge_index) for t in range(self.T)], dim=1)`

T could be big, graphs have have varying number of nodes.. so reshaping is tricky.. how about repeating the edge index? or any better ideas?

**Comment by rusty1s** [comment]
You can just permute `x` to `[T, batch_size * num_nodes, in_channels]` via `x = x.permute(1, 0, 2)`.

**Comment by engmubarak48** [comment]
This was very helpful and smart way to hand the extra dimension. Thank you @rusty1s :)
